pkg/notifier/notifier.go

```diff
@@ -0,0 +1,154 @@
+package notifier
+
+import (
+	"context"
+	"reflect"
+	"sync"
+)
+
+// Function signature used for Around subscriptions
+type AroundHandler = func(context.Context, func(ctx context.Context))
+
+// Function signature used for On subscriptions
+type OnHandler = func(context.Context)
+
+// The Notifiable interface is used to define Notifiers.
+type Notifiable interface {
+	Around(name string, handler AroundHandler)
+	On(name string, handler OnHandler)
+	RemoveOn(name string, handler OnHandler)
+	RemoveAround(name string, handler AroundHandler)
+	Emit(name string, ctx context.Context, f func(context.Context))
+}
+
+type nullNotifier struct{}
+
+var _ Notifiable = (*nullNotifier)(nil)
+
+// NullNotifier satisfies the Notifiable interface but does not persist Around
+// or On subscriptions.
+var NullNotifier = &nullNotifier{}
+
+func (n *nullNotifier) Around(name string, handler AroundHandler)       {}
+func (n *nullNotifier) On(name string, handler OnHandler)               {}
+func (n *nullNotifier) RemoveOn(name string, handler OnHandler)         {}
+func (n *nullNotifier) RemoveAround(name string, handler AroundHandler) {}
+func (n *nullNotifier) Emit(name string, ctx context.Context, handler func(context.Context)) {
+	handler(ctx)
+}
+
+// Notifier exposes hooks to subscribe and emit notifications that pass a
+// context.Context value allowing for easy implementation of custom logging,
+// observability, and other use-cases.
+type Notifier struct {
+	aroundSubscriptions map[string][]AroundHandler
+	onSubscriptions     map[string][]OnHandler
+
+	mu sync.Mutex
+}
+
+var _ Notifiable = (*Notifier)(nil)
+
+// New returns an empty Notifier.
+func New() *Notifier {
+	return &Notifier{
+		aroundSubscriptions: make(map[string][]AroundHandler),
+		onSubscriptions:     make(map[string][]OnHandler),
+	}
+}
+
+// Emit calls each subscription for the given name synchronously.
+//
+// Around subscriptions can pass a context to the provided callback that will
+// be passed to the next subscription if there is one, otherwise it is passed
+// to f.
+func (n *Notifier) Emit(name string, ctx context.Context, f func(ctx context.Context)) {
```

Comment:
Maybe it would be better to replace `name string` with `key interface{}` to follow the same pattern as `context`?

Reply:
That makes sense to me